### PR TITLE
Fix bugs in code that sets up higher half TSS/GDT

### PR DESCRIPTION
**Pull request by mpetch**

### CODE REVIEW

**Comment by IlanVinograd** [comment]
@mpetch 
so my bug was because I used init gdt inside `kernel_entry.s`?

**Comment by IlanVinograd** [comment]
and what is the differences between this `while (1) __asm__ __volatile__ ("hlt");` and  `while (1)` ?

**Comment by mpetch** [comment]
Mostly it was because of where you called init_gdt from but as well you set
ESP to the higher half stack to call functions before the higher half
addresses had been mapped. You can't use a `call` instruction with a
stack where memory hasn't been mapped or where memory may not exist. There
are also two *separate* stacks once `_start` begins executing. One is the
kernel stack and the other is the stack in the TSS to eventually
transition to ring 0 from user mode (ring 3).You don't have user mode code
yet so it isn't being used.

Your GDT/TSS code seemed pretty much fine except for setting up the esp0
stack.Your IDT also looks okay as well. I loaded your kernel in BOCHS and
did a `info gdt`, `info idt`, and `info tss` to ensure that was the case.

As for the `HLT` instruction in a loop. `while(1);` creates a loop that is
basically a jmp that keeps jumping to itself. That works but such a tight
loop keeps a CPU very busy (and it could run hotter). Usually people put a
`hlt` in a loop so that the processor waits until the next interrupt and
then loops back again. When interrupts are off it will basically sit there
forever (unless you get a non maskable interrupt).Once you enable
interrupts the HLT loop is much more efficient.

So you can remove the HLT but I recommend keeping it.

**Comment by IlanVinograd** [comment]
Oh, I got it. So now, I probably need to create stdlib.h for memory allocation, implement alloc_page() for paging, and then continue working on scheduling and processes or to fs and do FAT32.

**Comment by IlanVinograd** [comment]
Through which command did you check in Bochs for info gdt, info idt, and info tss?

**Comment by mpetch** [comment]
Those were the commands:

Display the GDT:
*info gdt*

Display the IDT:
*info idt*

Display TSS:
*info tss*

Display page mappings:
*info tab*

Display general registers
*registers*

Display segment registers:
*sreg*

Display CR registers:
*creg*
